pkg/registry/blobs.go

```diff
@@ -72,43 +160,126 @@ func (b *blobs) handle(resp http.ResponseWriter, req *http.Request) *regError {
 	digest := req.URL.Query().Get("digest")
 	contentRange := req.Header.Get("Content-Range")
 
+	repo := req.URL.Host + path.Join(elem[1:len(elem)-2]...)
+
 	switch req.Method {
 	case http.MethodHead:
-		b.lock.Lock()
-		defer b.lock.Unlock()
-		b, ok := b.contents[target]
-		if !ok {
+		h, err := v1.NewHash(target)
+		if err != nil {
 			return &regError{
-				Status:  http.StatusNotFound,
-				Code:    "BLOB_UNKNOWN",
-				Message: "Unknown blob",
+				Status:  http.StatusBadRequest,
+				Code:    "NAME_INVALID",
+				Message: "invalid digest",
 			}
 		}
 
-		resp.Header().Set("Content-Length", fmt.Sprint(len(b)))
-		resp.Header().Set("Docker-Content-Digest", target)
+		var size int64
+		if b.blobStatHandler != nil {
+			size, err = b.blobStatHandler.Stat(req.Context(), repo, h)
+			if errors.Is(err, errNotFound) {
+				return regErrBlobUnknown
+			} else if err != nil {
+				var rerr redirectError
+				if errors.As(err, &rerr) {
+					http.Redirect(resp, req, rerr.Location, rerr.Code)
+					return nil
+				}
+				return regErrInternal(err)
+			}
+		} else {
+			rc, err := b.blobHandler.Get(req.Context(), repo, h)
+			if errors.Is(err, errNotFound) {
+				return regErrBlobUnknown
+			} else if err != nil {
+				var rerr redirectError
+				if errors.As(err, &rerr) {
+					http.Redirect(resp, req, rerr.Location, rerr.Code)
+					return nil
+				}
+				return regErrInternal(err)
+			}
+			defer rc.Close()
+			size, err = io.Copy(ioutil.Discard, rc)
+			if err != nil {
+				return regErrInternal(err)
+			}
```

Comment:
Or we could require `Stat`, but provide a `partial.Blah` like method that "completes" the interface we expect.

Reply:
What we have now is effectively the same, without a new partial type. It might be worth adding to make implementors' lives easier, but I'll put that off for now until these interfaces have settled some more.